import error

Again, this error in CI:

```
        if maxBytes > 0 or backupCount > 0:
>           handler = logging.handlers.RotatingFileHandler(log_file, maxBytes=maxBytes, backupCount=backupCount)
E           AttributeError: module 'logging' has no attribute 'handlers'

../../../micromamba-root/envs/anaconda-test-env-py-3.9/lib/python3.9/site-packages/apstools/utils/log_utils.py:64: AttributeError
```

Yet the Python 3.9 documentation says that [`logging.handlers.RotatingFileHandler()`](https://docs.python.org/3.9/library/logging.handlers.html#rotatingfilehandler) is available.  So this might be an import problem (needing `import logging.handlers` in apstools upstream?)?

_Originally posted by @prjemian in https://github.com/BCDA-APS/bluesky_training/issues/83#issuecomment-1481997304_

--- a/CHANGES.rst
+++ b/CHANGES.rst
@@ -20,10 +20,24 @@ The project `milestones <https://github.com/BCDA-APS/apstools/milestones>`_
 describe the future plans.
 
 ..
-   1.6.15
+   1.6.16
    ******
 
    release expected by 2023-04-07
+
+1.6.15
+******
+
+release expected by 2023-03-27
+------------
+
+* add ``log_path=None`` kwarg to ``apstools.utils.stream_log_handler()``
+
+Fixes
+------------
+
+- resolved ``AttributeError`` for certain logging configurations
+
 
 1.6.14
 ******

--- a/apstools/utils/log_utils.py
+++ b/apstools/utils/log_utils.py
@@ -12,8 +12,12 @@ There is a guide describing :ref:`howto_session_logs`.
    ~stream_log_handler
 """
 
-import logging
 import pathlib
+from logging import DEBUG
+from logging import FileHandler
+from logging import Formatter
+from logging import StreamHandler
+from logging.handlers import RotatingFileHandler
 
 
 def file_log_handler(
@@ -58,16 +62,16 @@ def file_log_handler(
     """
     log_path = log_path or get_log_path()
     log_file = log_path / f"{file_name_base}.log"
-    level = level or logging.DEBUG
+    level = level or DEBUG
 
     if maxBytes > 0 or backupCount > 0:
-        handler = logging.handlers.RotatingFileHandler(log_file, maxBytes=maxBytes, backupCount=backupCount)
+        handler = RotatingFileHandler(log_file, maxBytes=maxBytes, backupCount=backupCount)
     else:
-        handler = logging.FileHandler(log_file)
+        handler = FileHandler(log_file)
 
     handler.setLevel(level)
 
-    formatter = logging.Formatter(
+    formatter = Formatter(
         (
             "|%(asctime)s"
             "|%(levelname)s"
@@ -96,7 +100,7 @@ def get_log_path():
     return path
 
 
-def setup_IPython_console_logging(logger=None, filename="ipython_console.log"):
+def setup_IPython_console_logging(logger=None, filename="ipython_console.log", log_path=None):
     """
     Record all input (``In``) and output (``Out``) from IPython console.
 
@@ -109,6 +113,10 @@ def setup_IPython_console_logging(logger=None, filename="ipython_console.log"):
         *str*:
         Name of the log file.
         (default: ``ipython_console.log``)
+    log_path : *str*
+        Directory to store the log file. Full name is
+        ``f"<log_path>/{file_name_base}.log"``.
+        default: (the present working directory)/LOG_DIR_BASE
     """
     try:
         from IPython import get_ipython
@@ -116,7 +124,8 @@ def setup_IPython_console_logging(logger=None, filename="ipython_console.log"):
         # start logging console to file
         # https://ipython.org/ipython-doc/3/interactive/magics.html#magic-logstart
         _ipython = get_ipython()
-        io_file = get_log_path() / filename
+        log_path = pathlib.Path(log_path or get_log_path())
+        io_file = log_path / filename
         if _ipython is not None:
             _ipython.magic(f"logstart -o -t {io_file} rotate")
         if logger is not None:
@@ -142,13 +151,13 @@ def stream_log_handler(formatter=None, level="INFO"):
         Name of the logging level to report.
         (default: ``INFO``)
     """
-    handler = logging.StreamHandler()
+    handler = StreamHandler()
 
     # nice output format
     # https://docs.python.org/3/library/logging.html#logrecord-attributes
     if formatter is None:
         # fmt: off
-        formatter = logging.Formatter(
+        formatter = Formatter(
             (
                 "%(levelname)-.1s"		# only first letter
                 " %(asctime)s"
@@ -163,6 +172,7 @@ def stream_log_handler(formatter=None, level="INFO"):
     handler.setLevel(level)
 
     return handler
+
 
 # -----------------------------------------------------------------------------
 # :author:    Pete R. Jemian

--- a/apstools/utils/tests/test_log_utils.py
+++ b/apstools/utils/tests/test_log_utils.py
@@ -25,29 +25,58 @@ def test_get_log_path(tempdir):
     assert log_path.exists()
 
 
-def test_file_log_handler(tempdir):
+@pytest.mark.parametrize(
+    "maxBytes, backupCount, log_path, level",
+    [
+        [0, 0, None, None],  # defaults
+        [1_000_000, 9, None, None],  # training repo uses this
+        [0, 9, None, None],
+        [1_000_000, 0, None, None],
+        [1_000_000, 0, None, "DEBUG"],
+        [0, 0, "/tmp", None],
+    ],
+)
+def test_file_log_handler(maxBytes, backupCount, log_path, level, tempdir):
     pathlib.os.chdir(tempdir)
-    handler = file_log_handler("my_log_file")
-    assert isinstance(handler, logging.Handler)
-
-    log_path = get_log_path()
+    log_path = pathlib.Path(log_path or get_log_path())
     assert log_path.exists()
+
+    handler = file_log_handler(
+        "my_log_file",
+        maxBytes=maxBytes,
+        backupCount=backupCount,
+        log_path=log_path,
+        level=level,
+    )
+    assert isinstance(handler, logging.Handler)
 
     logfile = log_path / "my_log_file.log"
     assert logfile.exists()
 
-    # TODO: more tests
 
-
-def test_stream_log_handler():
-    handler = stream_log_handler()
+@pytest.mark.parametrize(
+    "formatter, level",
+    [
+        [None, "INFO"],  # defaults
+        [None, "DEBUG"],
+    ],
+)
+def test_stream_log_handler(formatter, level):
+    handler = stream_log_handler(formatter=formatter, level=level)
     assert isinstance(handler, logging.Handler)
     # TODO: more tests
 
 
-def test_setup_IPython_console_logging(tempdir):
+@pytest.mark.parametrize(
+    "logger, filename, log_path",
+    [
+        [None, "ipython_console.log", None],  # defaults
+        [None, "ipython_console.log", "/tmp"],
+    ],
+)
+def test_setup_IPython_console_logging(logger, filename, log_path, tempdir):
     pathlib.os.chdir(tempdir)
-    setup_IPython_console_logging()
+    setup_IPython_console_logging(logger=logger, filename=filename, log_path=log_path)
 
     log_path = get_log_path()
     assert log_path.exists()

